src/neon/neon_utils.rs

```diff
@@ -0,0 +1,278 @@
+    vreinterpretq_f32_f64(vtrn2q_f64(
+        vreinterpretq_f64_f32(left),
+        vreinterpretq_f64_f32(right),
+    ))
+}
+
+// Pack low (1st) and high (2nd) complex
+// left: r1.re, r1.im, r2.re, r2.im
+// right: l1.re, l1.im, l2.re, l2.im
+// --> r1.re, r1.im, l2.re, l2.im
+#[inline(always)]
+pub unsafe fn extract_lo_hi_f32(left: float32x4_t, right: float32x4_t) -> float32x4_t {
+    vcombine_f32(vget_low_f32(left), vget_high_f32(right))
+}
+
+// Pack  high (2nd) and low (1st) complex
+// left: r1.re, r1.im, r2.re, r2.im
+// right: l1.re, l1.im, l2.re, l2.im
+// --> r2.re, r2.im, l1.re, l1.im
+#[inline(always)]
+pub unsafe fn extract_hi_lo_f32(left: float32x4_t, right: float32x4_t) -> float32x4_t {
+    vcombine_f32(vget_high_f32(left), vget_low_f32(right))
+}
+
+// Reverse complex
+// values: a.re, a.im, b.re, b.im
+// --> b.re, b.im, a.re, a.im
+#[inline(always)]
+pub unsafe fn reverse_complex_elements_f32(values: float32x4_t) -> float32x4_t {
+    vcombine_f32(vget_high_f32(values), vget_low_f32(values))
+}
+
+// Reverse complex and then negate hi complex
+// values: a.re, a.im, b.re, b.im
+// --> b.re, b.im, -a.re, -a.im
+#[inline(always)]
+pub unsafe fn reverse_complex_and_negate_hi_f32(values: float32x4_t) -> float32x4_t {
+    vcombine_f32(vget_high_f32(values), vneg_f32(vget_low_f32(values)))
+}
+
+// Invert sign of high (2nd) complex
+// values: a.re, a.im, b.re, b.im
+// -->  a.re, a.im, -b.re, -b.im
+//#[inline(always)]
+//pub unsafe fn negate_hi_f32(values: float32x4_t) -> float32x4_t {
+//    vcombine_f32(vget_low_f32(values), vneg_f32(vget_high_f32(values)))
+//}
+
+// Duplicate low (1st) complex
+// values: a.re, a.im, b.re, b.im
+// --> a.re, a.im, a.re, a.im
+#[inline(always)]
+pub unsafe fn duplicate_lo_f32(values: float32x4_t) -> float32x4_t {
+    vreinterpretq_f32_f64(vtrn1q_f64(
+        vreinterpretq_f64_f32(values),
+        vreinterpretq_f64_f32(values),
+    ))
+}
+
+// Duplicate high (2nd) complex
+// values: a.re, a.im, b.re, b.im
+// --> b.re, b.im, b.re, b.im
+#[inline(always)]
+pub unsafe fn duplicate_hi_f32(values: float32x4_t) -> float32x4_t {
+    vreinterpretq_f32_f64(vtrn2q_f64(
+        vreinterpretq_f64_f32(values),
+        vreinterpretq_f64_f32(values),
+    ))
+}
+
+// transpose a 2x2 complex matrix given as [x0, x1], [x2, x3]
+// result is [x0, x2], [x1, x3]
+#[inline(always)]
+pub unsafe fn transpose_complex_2x2_f32(left: float32x4_t, right: float32x4_t) -> [float32x4_t; 2] {
+    let temp02 = extract_lo_lo_f32(left, right);
+    let temp13 = extract_hi_hi_f32(left, right);
+    [temp02, temp13]
+}
+
+// Complex multiplication.
+// Each input contains two complex values, which are multiplied in parallel.
+#[inline(always)]
+pub unsafe fn mul_complex_f32(left: float32x4_t, right: float32x4_t) -> float32x4_t {
+    // ARMv8.2-A introduced vcmulq_f32 and vcmlaq_f32 for complex multiplication, these intrinsics are not yet available.
```

Comment:
Is this planned to be included into rust at some point? If it's gonna be soon-ish, maybe it'll be worth it to wait for it to be stabilized, and/or push for it to be stabilized.

Reply:
I have looked a bit more at these instructions after writing that. They are a quite recent addition from (IIRC) ARMv8.4. So there aren't that many arm chips out there that have them (most are ARMv8.2). It might be better to leave these functions like they are.
I don't know if anyone is planning on adding them to Rust.

Reply:
@HEnquist `vcmlaq_f32` is on the radar as it is part of the [Arm C Language Extensions (ACLE) specification](https://developer.arm.com/documentation/ihi0073/latest), see the linked Google spreadsheet in https://github.com/rust-lang/stdarch/issues/148#issuecomment-792691514 and will be added to stdarch eventually. Isn't `vcmulq_f32` MVE-only?

Reply:
Alright! `vcmlaq` and `vcadd` should be quite useful. The `vcmul` variants seem to be MVE-only yes. 

I have been looking for CPUs have them, not so easy to find. The only relevant one (meaning one people can get their hands on) at the moment seems to be the Apple M1.